3dinput

wenbinc-Bin opened the pull request:
Add input frame from and to vulkan for ffmpeg-vaapi and ffmpeg-qsv.

BNieuwenhuizen committed:
hwcontext_vaapi: Use PRIME_2 memory type for modifiers.

This way we can pass explicit modifiers in. Sometimes the
modifier matters for the number of memory planes that
libva accepts, in particular when dealing with
driver-compressed textures. Furthermore the driver might
not actually be able to determine the implicit modifier
if all the buffer-passing has used explicit modifier.
All these issues should be resolved by passing in the
modifier, and for that we switch to using the PRIME_2
memory type.

Tested with experimental radeonsi patches for modifiers
and kmsgrab. Also tested with radeonsi without the
patches to double-check it works without PRIME_2 support.

v2:
  Cache PRIME_2 support to avoid doing two calls every time on
  libva drivers that do not support it.

v3:
  Remove prime2_vas usage.

wenbinc-Bin committed:
libavutil/hwcontext_vulkan: Add single_linear flag to make vulkan compatible with vaapi device.

Vaapi can import external surface, but all the planes of the external
frames should be in the same drm object. A new flag is introduced and
vulkan can choose to allocate planes in one memory according this flag.
This flag will be enabled when the vulkan device is derived from vaapi
device, so that this change will not affect current vulkan behaviour.

Signed-off-by: Wenbin,Chen <wenbin.chen@intel.com>

wenbinc-Bin committed:
libavutil/hwcontext_vulkan: Allocate vkFrame in one memory

The vaapi can import external frame, but the planes of the external
frames should be in the same drm object. I add a new function to
allocate vkFrame in one memory and vulkan device will choose a way
to allocate memory by signgle_linear flag.
A new variable is added to AVVKFrame to store the offset of each plane.

Now the following command line works:

ffmpeg -hwaccel vaapi -hwaccel_device /dev/dri/renderD128 -hwaccel_output_format
vaapi -i input_1080p.264 -vf "hwmap=derive_device=vulkan,format=vulkan,
scale_vulkan=1920:1080,hwmap=derive_device=vaapi,format=vaapi" -c:v h264_vaapi output.264

Signed-off-by: Wenbin,Chen <wenbin.chen@intel.com>

wenbinc-Bin committed:
libavutil/hwcontext_vaapi: Add a new nv12 format map to support vulkan frame

Vulkan will map nv12 to R8 and GR88, so add this map to vaapi to support
vulkan frame.

Signed-off-by: Wenbin,Chen <wenbin.chen@intel.com>

wenbinc-Bin committed:
libavutil/hwcontext_vaapi: Add support to derive vaapi device from qsv

Add support to derive vaapi device_context from qsv device_context

Now the following command line works:

ffmpeg -v verbose -hwaccel qsv -c:v h264_qsv -i input_1080p.264 -vf
"hwmap=derive_device=vaapi,format=vaapi,hwmap=derive_device=vulkan,
format=vulkan,scale_vulkan=w=1920:h=1080,hwdownload,format=nv12" -f rawvideo output.yuv

Signed-off-by: Wenbin,Chen <wenbin.chen@intel.com>

wenbinc-Bin committed:
libavcodec/qsvenc: Add support to qsv to encode external surface.

Qsv encoder only encode the frame that are pre-allocated, so qsv encoder
cannot encode the frame mapped from vulkan. In fact, MediaSDK can encode
frame that are dynamically created. I add the support for qsv to encode
external frame. The resource will be allocated and released in real time
if the encoded frame is not in the frame pool.

Now the following command line works:

ffmpeg -v verbose -hwaccel qsv -c:v h264_qsv -i input_1080p.264 -vf
"hwmap=derive_device=vaapi,format=vaapi,hwmap=derive_device=vulkan,
format=vulkan,scale_vulkan=w=1920:h=1080,hwmap=derive_device=vaapi,format=vaapi,
hwmap=derive_device=qsv,format=qsv" -c:v h264_qsv output.264

Signed-off-by: Wenbin,Chen <wenbin.chen@intel.com>

wenbinc-Bin committed:
libavutil/hwcontext_vulkan: Add hwupload and hwdownload support when using single_linear flag.

Add hwupload and hwdownload support to vulkan when frames are allocated
in one memory

Signed-off-by: Wenbin,Chen <wenbin.chen@intel.com>

wenbinc-Bin commented:
@xuguangxin @xhaihao @feiwan1 

Could you please help me to review these patches.

Thanks
Wenbin

wenbinc-Bin commented:
@xuguangxin @xhaihao @feiwan1 
Is there any feedback on this patchset?
Thank you

xuguangxin commented on libavutil/hwcontext_vulkan.c:
```diff
@@ -1349,7 +1349,7 @@ static int vulkan_device_init(AVHWDeviceContext *ctx)
         }
     }
 
-    err = load_functions(ctx, 1, 0);
```
why the original author, not use 1?
any side effects for this change?

wenbinc-Bin replied:
I don't know, author didn't mention in his patch. But without using 1, there will be an segmentation error because vk->CreateCommandPool is not load. 
Segmentation error occur in this line:
https://github.com/FFmpeg/FFmpeg/blob/2fac1e370c9cf48b2e9d4a7f2c0d7236017f9bbe/libavutil/hwcontext_vulkan.c#L1018

xuguangxin replied:
Without any of you patch, it will not have a crash, right?

wenbinc-Bin replied:
without this patch it will crash. without other pathes, the command line will report error.

wenbinc-Bin replied:
I submit a ticket. https://trac.ffmpeg.org/ticket/9279#ticket.
Temporarily use this patch before they fix the issue.

xuguangxin commented on libavutil/hwcontext_vulkan.c:
```diff
@@ -1448,6 +1458,8 @@ static int vulkan_device_derive(AVHWDeviceContext *ctx,
         if (strstr(vendor, "AMD"))
             dev_select.vendor_id = 0x1002;
 
+        av_dict_set_int(&opts, "single_linear", 1, NULL);
```
will this impact AMD? could we just use single_linear for intel only?

wenbinc-Bin replied:
Good advice, and now change it.

xuguangxin commented on libavutil/hwcontext_vulkan.c:
```diff
@@ -1659,6 +1659,88 @@ static void vulkan_frame_free(void *opaque, uint8_t *data)
     av_free(f);
 }
 
+static int alloc_bind_single_mem(AVHWFramesContext *hwfc, AVVkFrame *f,
```
nearly same code as alloc_bind_mem, could we pass one flag to 
static int alloc_bind_mem(AVHWFramesContext *hwfc, AVVkFrame *f,    void *alloc_pnext, size_t alloc_pnext_stride, int use_single_linear_images )

wenbinc-Bin replied:
I've merged them into one function.

xuguangxin commented on libavutil/hwcontext_vulkan.c:
```diff
@@ -2258,16 +2258,31 @@ static int vulkan_map_frame_to_mem(AVHWFramesContext *hwfc, AVFrame *dst,
             goto fail;
         }
-        mapped_mem_count++;
+        for (int i = 0; i < planes; i++) {
```
missed mapped_mem_count?
How about 
```
loop = use_single_linear_images ? 1 :planes

for (int i = 0; i < loop; i++)
    xxxx..
if (use_single_linear_images) {
   for (int i = i; i < planes; i++) {
         dst->data[i] = dst->data[0]+ f->offset[i];
   }
}
```

wenbinc-Bin replied:
Good advice, I've changed it.

xuguangxin commented on libavutil/hwcontext_vulkan.c:
```diff
@@ -1659,6 +1659,88 @@ static void vulkan_frame_free(void *opaque, uint8_t *data)
+            req.memoryRequirements.size = FFALIGN(req.memoryRequirements.size,
+                                                  p->props.properties.limits.minMemoryMapAlignment);
+
+        mem_size_list[i] += FFALIGN(req.memoryRequirements.size/hwfc->width, 64) * hwfc->width;
```
Where is the 64 come from? in the future platforms it may change.
Could we get the 64 from somewhere?

wenbinc-Bin replied:
Here I made a mistake, I want the offset to be multiple of width.

xuguangxin commented on fftools/ffmpeg_qsv.c:
```diff
@@ -48,6 +48,22 @@ static int qsv_device_init(InputStream *ist)
+                err = AVERROR(ENOMEM);
+            else
+                err = 0;
+            goto err_out;
```
why we need err_out for err = 0

wenbinc-Bin replied:
Because "err_out" release the resource, I want to make sure the resource is released. However, I found at this line, the resource is not allocated yet, so it is unnecessary. Now I changed it. Thanks.

xuguangxin commented on libavcodec/qsvenc.c:
```diff
@@ -1304,12 +1312,18 @@ static int submit_frame(QSVEncContext *q, const AVFrame *frame,
+            if (ret >= 0)
+                qf->surface.Data.MemId = &q->frames_ctx.mids[ret];
+        }
+        if (!q->frames_ctx.mids || ret < 0) {
```
if q->frames_ctx.mids is not NULL, ret < 0 is a error.

wenbinc-Bin replied:
if mids is not NULL, ret is the return value from ff_qsv_find_surface_idx(), so if ret >=0 we skip this branch, and if ret < 0, it means the surface is not in the pool, and we need go into this branch to create new resource for this frame.

xuguangxin replied:
in some error cases, mids are not null, and ret will < 0 too. In this case we can't create a new resource for it, we need to report an error.

wenbinc-Bin replied:
Ok, I got your point.
The ff_qsv_find_surface_idx() is a simple function, It returns minus value only when the surface is not found, so I don't category the returned value.

xuguangxin commented on libavcodec/qsvenc.c:
```diff
@@ -1304,12 +1312,18 @@ static int submit_frame(QSVEncContext *q, const AVFrame *frame,
+        }
+        if (!q->frames_ctx.mids || ret < 0) {
+            QSVMid *mid = NULL;
+            mid = (QSVMid *)av_mallocz(sizeof(*mid));
```
need check mid

wenbinc-Bin replied:
Add check, thanks.

xuguangxin commented on libavutil/hwcontext_qsv.c:
```diff
@@ -1044,9 +1046,14 @@ static int qsv_map_to(AVHWFramesContext *dst_ctx,
-        av_log(dst_ctx, AV_LOG_ERROR, "Trying to map from a surface which "
-               "is not in the mapped frames context.\n");
-        return AVERROR(EINVAL);
+        new_sur = (mfxFrameSurface1 *)av_mallocz(sizeof(*new_sur));
```
need check the malloc

wenbinc-Bin replied:
Add check, thanks.

xuguangxin commented on libavutil/hwcontext_vulkan.c:
```diff
@@ -2188,6 +2188,7 @@ static int vulkan_map_frame_to_mem(AVHWFramesContext *hwfc, AVFrame *dst,
 {
     VkResult ret;
     int err, mapped_mem_count = 0;
+    int loop = 0;
```
init is not needed

wenbinc-Bin replied:
change it, thanks.

xuguangxin commented on fftools/ffmpeg_qsv.c:
```diff
@@ -48,6 +48,21 @@ static int qsv_device_init(InputStream *ist)
+
+    if (ist->hwaccel_device) {
+        hw_device = hw_device_get_by_name(ist->hwaccel_device);
+        if (hw_device && hw_device->type == AV_HWDEVICE_TYPE_QSV) {
```
maye you can use
if (hw_device) {
 if (hw_device->type != ..)


}

wenbinc-Bin replied:
change it, thanks

xuguangxin commented on fftools/ffmpeg_qsv.c:
```diff
@@ -48,6 +48,21 @@ static int qsv_device_init(InputStream *ist)
+            hw_device_ctx = av_buffer_ref(hw_device->device_ref);
+            if (!hw_device_ctx)
+                return AVERROR(ENOMEM);
+            else
```
else is not needed.

wenbinc-Bin replied:
change it, thanks

xuguangxin commented on fftools/ffmpeg_qsv.c:
```diff
@@ -48,6 +48,21 @@ static int qsv_device_init(InputStream *ist)
+    if (ist->hwaccel_device) {
+        hw_device = hw_device_get_by_name(ist->hwaccel_device);
+        if (hw_device && hw_device->type == AV_HWDEVICE_TYPE_QSV) {
+            hw_device_ctx = av_buffer_ref(hw_device->device_ref);
```
who will release this?

wenbinc-Bin replied:
Add unref operation in qsv_uninit(). It miss an unref even without this patchset. Seems it is a bug for ffmpeg-qsv.

xuguangxin commented on libavutil/hwcontext_vulkan.c:
```diff
@@ -1696,6 +1699,23 @@ static int alloc_bind_mem(AVHWFramesContext *hwfc, AVVkFrame *f,
+                return AVERROR(EINVAL);
+            }
+
+            mem_size_list[i] += (req.memoryRequirements.size/hwfc->width + 1) * hwfc->width;
```
why not use req.memoryRequirements.size?
is this workable for other format like i420?

wenbinc-Bin replied:
Because req.memoryRequirements.size is not perfectly aligned with width, and libva require that.
It should work for other format as well and I will test it.

xuguangxin commented on libavutil/hwcontext_qsv.c:
```diff
@@ -1044,9 +1046,20 @@ static int qsv_map_to(AVHWFramesContext *dst_ctx,
+            return AVERROR(ENOMEM);
+        new_vasur = (VASurfaceID *)av_mallocz(sizeof(VASurfaceID));
+        if (!new_vasur)
+            return AVERROR(ENOMEM);
```
this return will leak new_sur

wenbinc-Bin replied:
change it, thanks.

wenbinc-Bin committed:
libavutil/hwcontext_qsv: Fix an issue when no hw_frame is allocated

Some filters and codecs use surfaces[0] to get frameInfo, but there is not
guarantee that the surfaces is allocated. Add a variable to store the
surface information, and filter and codec and get mfxSurface information
from it.

Now the following command can work:

ffmpeg -v verbose -hwaccel qsv -c:v h264_qsv -i 1080p.264 -vf
"hwmap=derive_device=vaapi,format=vaapi,hwmap=derive_device=vulkan,format=vulkan,
scale_vulkan=w=1920:h=1080,hwmap=derive_device=vaapi,format=vaapi,hwmap=derive_device=qsv,
format=qsv,scale_qsv=w=720:h=480" -c:v h264_qsv output.264

Signed-off-by: Wenbin,Chen <wenbin.chen@intel.com>

xhaihao reviewed (commented):
Someone disagree with the change of hwaccel_device, see http://ffmpeg.org/pipermail/ffmpeg-devel/2020-September/269074.html for detail

wenbinc-Bin commented:
> Someone disagree with the change of hwaccel_device, see http://ffmpeg.org/pipermail/ffmpeg-devel/2020-September/269074.html for detail

I read the thread line...  According to softworkz's first reply, he seems to agree to use -hwaccel_device to refer to an existing device created with -init_hw_device by name. Maybe he can accept this patch?

I see mypopy disagree to mix qsv and vaapi in one ffmpeg pipeline, which is what I am going to do...

xhaihao commented:
> > Someone disagree with the change of hwaccel_device, see http://ffmpeg.org/pipermail/ffmpeg-devel/2020-September/269074.html for detail
> 
> I read the thread line... According to softworkz's first reply, he seems to agree to use -hwaccel_device to refer to an existing device created with -init_hw_device by name. Maybe he can accept this patch?
> 

According to the feedback in this and other threads, we should support HW_DEVICE_CTX and remove AD_HOC for qsv decoder. If so, the device created by init_hw_device can be used for qsv decoder, like as qsv encoder. I am looking into it. Does it block your implementation for "vaapi decoding + qsv encoding" ? 

> I see mypopy disagree to mix qsv and vaapi in one ffmpeg pipeline, which is what I am going to do...

Yes, he disagreed with the combination of vaapi and qsv. You may submit your patches to see whether there are more comments.

wenbinc-Bin commented:
> According to the feedback in this and other threads, we should support HW_DEVICE_CTX and remove AD_HOC for qsv decoder. If so, the device created by init_hw_device can be used for qsv decoder, like as qsv encoder. I am looking into it. Does it block your implementation for "vaapi decoding + qsv encoding" ?

I think it doesn't block me.
> 
> > I see mypopy disagree to mix qsv and vaapi in one ffmpeg pipeline, which is what I am going to do...
> 
> Yes, he disagreed with the combination of vaapi and qsv. You may submit your patches to see whether there are more comments.

Ok, thanks Haihao.

wenbinc-Bin commented:
I see vulkan init error is already fixed, so I remove one of my patches.

wenbinc-Bin commented:
After Haihao mentioned the discussion about -hwaccel_device, I check the code and found that the using "-hwaccel_device qs" to choose the device_ctx for -hwaccel qsv may not comply the logic of ffmpeg_hw, so I remove that patch and add a new patch to support deriving vaapi device_ctx from qsv to implemnet 3Dinput function.

wenbinc-Bin commented:
On DG2 libva is not compatible with vulkan. Will reopen this pr when this problem is fixed on driver side.
